File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/layout/flexbox/FlexboxExtractor.swift
//
//  FlexboxExtractor.swift
//  StyleEngine/layout/flexbox — Phase 7 step 2 (flexbox).
//
//  Folds the 11 flexbox-family longhands (plus BoxOrient, skipped below)
//  into the shared LayoutAggregate. The aggregate is consumed once by
//  LayoutApplier / ComponentRenderer to pick the SwiftUI container kind
//  (HStack / VStack / ZStack / FlowLayout) and the per-child modifier
//  chain — see LayoutAggregate.swift for the rationale on why these are
//  fused rather than one-modifier-per-property.
//
//  Scope (task: pedantic-bhabha, Phase 7b flexbox slice):
//    Display, FlexDirection, FlexWrap,
//    FlexGrow, FlexShrink, FlexBasis,
//    JustifyContent, AlignItems, AlignContent, AlignSelf, Order.
//  BoxOrient: skipped — parser is not registered per Phase 7 fixtures
//  README, so the property never reaches the iOS runtime in practice.
//

// Foundation is enough — no SwiftUI types escape the aggregate.
import Foundation

enum FlexboxExtractor {

    /// Single linear pass. Later occurrences of the same property win —
    /// matches the CSS last-wins cascade within a single rule, same
    /// convention every other extractor in this tree follows.
    /// Returns the aggregate unchanged when none of the 11 properties
    /// are present; the caller (`LayoutExtractor`) decides whether the
    /// aggregate was `touched`.
    static func extract(from properties: [IRProperty], into agg: inout LayoutAggregate) {
        // Walk once; dispatch by property type string. Unknown names in
        // the flex family (BoxOrient today) are silently ignored to keep
        // the extractor honest about what it actually supports.
        for prop in properties {
            switch prop.type {
            case "Display":
                // IR shape: plain string "FLEX" / "GRID" / "BLOCK" / …
                // or the parser-shell object {keyword: …}. `extractKeyword`
                // handles both; `normalize` uppercases + hyphen→underscore.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.display = mapDisplay(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "FlexDirection":
                // IR shape: plain string "ROW" / "ROW_REVERSE" / "COLUMN"
                // / "COLUMN_REVERSE". See FlexDirectionPropertyParser.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.flexDirection = mapFlexDirection(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "FlexWrap":
                // IR shape: plain string "NOWRAP" / "WRAP" / "WRAP_REVERSE".
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.flexWrap = mapFlexWrap(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "JustifyContent":
                // IR shape: keyword. Covers flex-start / flex-end / center
                // / space-between / space-around / space-evenly + CSS
                // logical start/end aliases.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.justifyContent = mapAlignment(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "AlignItems":
                // IR shape: keyword. Covers stretch / flex-start / flex-end
                // / center / baseline.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.alignItems = mapAlignment(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "AlignContent":
                // IR shape: keyword. Superset of AlignItems plus
                // space-between / around / evenly for multi-line packing.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.alignContent = mapAlignment(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "AlignSelf":
                // IR shape: keyword including "AUTO" for "inherit from
                // parent's align-items". We record auto so the applier
                // can skip emitting an override.
                if let kw = ValueExtractors.extractKeyword(prop.data) {
                    agg.alignSelf = mapAlignment(ValueExtractors.normalize(kw))
                    agg.touched = true
                }
            case "FlexGrow":
                // IR shape: {value: {type: "…Number", value: N}, normalizedValue: N}.
                // Reading normalizedValue is cheapest — the parser already
                // clamped negatives to 0 per the CSS spec.
                if let n = extractNormalizedNumber(prop.data) {
                    agg.flexGrow = n
                    agg.touched = true
                }
            case "FlexShrink":
                // Same IR shape as FlexGrow. SwiftUI has no analogue (noted
                // in the applier); we still fold so downstream selftests
                // can inspect the value.
                if let n = extractNormalizedNumber(prop.data) {
                    agg.flexShrink = n
                    agg.touched = true
                }
            case "FlexBasis":
                // IR shapes observed in fixtures:
                //   • plain string "auto" / "content"
                //   • {value: {px: N}, normalizedPixels: N} (absolute length)
                //   • {keyword: "auto"} (parser-shell fallback)
                if let v = extractFlexBasis(prop.data) {
                    agg.flexBasis = v
                    agg.touched = true
                }
            case "Order":
                // IR shape: raw integer (observed in fixtures). Fall back
                // to ValueExtractors.extractInt which also handles boxed
                // {value: N} shapes that other parsers occasionally emit.
                if let i = ValueExtractors.extractInt(prop.data) {
                    agg.order = i
                    agg.touched = true
                }
            case "BoxOrient":
                // Out of scope per task brief — parser not registered.
                // Leaving the case explicit so grep-coverage stays honest.
                break
            default:
                // Not a flexbox property — other extractors own it.
                continue
            }
        }
    }

    // MARK: - Keyword → enum maps

    /// Maps normalized (upper/underscore) CSS `display` keywords to the
    /// subset SwiftUI can represent. Unknown values fall back to `.block`
    /// because that's the CSS default for initial value of `display`.
    private static func mapDisplay(_ kw: String) -> DisplayKeyword {
        switch kw {
        // `inline-flex` renders the same as `flex` in our SDUI runtime —
        // we don't model inline run context for flex containers.
        case "FLEX", "INLINE_FLEX":     return .flex
        case "GRID", "INLINE_GRID":     return .grid
        case "INLINE", "INLINE_BLOCK":  return .inline
        case "NONE":                    return .none
        case "CONTENTS":                return .contents
        default:                        return .block
        }
    }

    /// Maps `flex-direction` keyword. Unknown → row (CSS initial value).
    private static func mapFlexDirection(_ kw: String) -> FlexDirectionKeyword {
        switch kw {
        case "ROW_REVERSE":    return .rowReverse
        case "COLUMN":         return .column
        case "COLUMN_REVERSE": return .columnReverse
        default:               return .row
        }
    }

    /// Maps `flex-wrap` keyword. Unknown → nowrap (CSS initial value).
    private static func mapFlexWrap(_ kw: String) -> FlexWrapKeyword {
        switch kw {
        case "WRAP":         return .wrap
        case "WRAP_REVERSE": return .wrapReverse
        default:             return .nowrap
        }
    }

    /// Maps the common CSS alignment keyword space used by justify-content
    /// / align-items / align-content / align-self. Unknown → normal so the
    /// applier can treat it as "no override".
    private static func mapAlignment(_ kw: String) -> AlignmentKeyword {
        switch kw {
        // Start / end family + physical aliases the parser normalises to
        // keyword strings. `LEFT`/`RIGHT` collapse into start/end because
        // SwiftUI stacks don't have separate physical alignments.
        case "FLEX_START", "START", "LEFT":   return .start
        case "FLEX_END",   "END",   "RIGHT":  return .end
        case "CENTER":                        return .center
        case "STRETCH":                       return .stretch
        case "BASELINE", "FIRST_BASELINE", "LAST_BASELINE":
            return .baseline
        case "SPACE_BETWEEN":                 return .spaceBetween
        case "SPACE_AROUND":                  return .spaceAround
        case "SPACE_EVENLY":                  return .spaceEvenly
        case "SELF_START":                    return .selfStart
        case "SELF_END":                      return .selfEnd
        case "AUTO":                          return .auto
        case "NORMAL":                        return .normal
        // CSS Anchor Positioning Level 1 §6: `anchor-center` collapses to
        // `center` whenever no default anchor is in scope. SwiftUI has no
        // anchor-positioning runtime so this fold is always correct here.
        // Mirrors Android's parseAlignment + Web's AlignSelfApplier.
        case "ANCHOR_CENTER":                 return .center
        default:                              return .normal
        }
    }

    // MARK: - Numeric / length helpers

    /// FlexGrow/FlexShrink IR: `{value: {value: N}, normalizedValue: N}`.
    /// Prefer `normalizedValue` when present (cheapest path), fall back
    /// to the nested `value.value` tuple, then to a raw number.
    private static func extractNormalizedNumber(_ v: IRValue) -> Double? {
        if case .object(let o) = v {
            if let n = o["normalizedValue"]?.doubleValue { return n }
            if let inner = o["value"]?.objectValue,
               let n = inner["value"]?.doubleValue { return n }
        }
        // Raw-number fallback covers the "we parsed it as a bare number"
        // path some shorthand expansions use.
        switch v {
        case .double(let d): return d
        case .int(let i):    return Double(i)
        default:             return nil
        }
    }

    /// FlexBasis IR shapes: plain "auto"/"content" strings, or
    /// `{value: {px: N}, normalizedPixels: N}` for lengths, or
    /// `{keyword: "auto"}` via the parser-shell fallback.
    private static func extractFlexBasis(_ v: IRValue) -> FlexBasisValue? {
        // String-form shortcut — fixtures use this for `auto` and `content`.
        if case .string(let s) = v {
            switch s.lowercased() {
            case "auto":    return .auto
            case "content": return .content
            default:        return nil
            }
        }
        // Object form — check for length first, then keyword.
        if case .object(let o) = v {
            // Normalized-pixels is populated for absolute lengths.
            if let n = o["normalizedPixels"]?.doubleValue {
                return .px(CGFloat(n))
            }
            // Nested length carrier: {value: {px: N}}.
            if let inner = o["value"]?.objectValue,
               let px = inner["px"]?.doubleValue {
                return .px(CGFloat(px))
            }
            // Direct {px: N} shape (rare but safe to handle).
            if let px = o["px"]?.doubleValue {
                return .px(CGFloat(px))
            }
            // Keyword parser-shell fallback.
            if let kw = o["keyword"]?.stringValue {
                switch kw.lowercased() {
                case "auto":    return .auto
                case "content": return .content
                default:        return nil
                }
            }
        }
        return nil
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/layout/flexbox/FlexboxExtractor.kt
package com.styleconverter.runtime.layout.flexbox

// Phase 7 step 2 — flexbox sub-extractor for the style-engine LayoutConfig.
//
// This extractor is a PARTIAL LayoutConfig producer: it consumes the IR
// property pair list and returns a LayoutConfig populated only in the
// flexbox fields (display, flex*, justify/align*, order). All grid/position/
// advanced/root fields are left null so the later phase-step extractors can
// merge into the same LayoutConfig without overwriting each other.
//
// Precedent:
//   - BorderSideExtractor — single-file multi-property extractor (27 props)
//   - FlexExtractor (sibling file) — the LEGACY extractor whose keyword
//     mappings we deliberately mirror one-for-one so ComponentRenderer
//     behaviour stays byte-identical when it's wired to the engine path.
//
// Registration: LayoutExtractor already claims every flexbox property name
// against PropertyRegistry in its init {} block. FlexboxExtractor must NOT
// re-register — doing so would trip the "duplicate registration" check.

import com.styleconverter.runtime.core.types.ValueExtractors
import com.styleconverter.runtime.layout.AlignmentKeyword
import com.styleconverter.runtime.layout.DisplayKind
import com.styleconverter.runtime.layout.FlexBasisValue
import com.styleconverter.runtime.layout.FlexDirection as EngineFlexDirection
import com.styleconverter.runtime.layout.FlexWrap as EngineFlexWrap
import com.styleconverter.runtime.layout.LayoutConfig
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.floatOrNull
import kotlinx.serialization.json.intOrNull

/**
 * Extract flexbox-only fields from an IR property stream into a [LayoutConfig].
 *
 * The 12 properties handled here match the Phase 7b scope list:
 *   Display, FlexDirection, FlexWrap,
 *   FlexGrow, FlexShrink, FlexBasis,
 *   JustifyContent, AlignItems, AlignContent,
 *   AlignSelf, Order,
 *   BoxOrient (legacy -webkit-box-orient, treated as FlexDirection).
 *
 * Every field on the returned [LayoutConfig] that is NOT flexbox-related is
 * null. Later sub-extractors (grid, position, advanced, root) are expected
 * to fill the remaining slots via `.copy(...)`. The merge strategy lives in
 * [com.styleconverter.runtime.layout.LayoutExtractor.extractLayoutConfig].
 */
object FlexboxExtractor {

    /**
     * Build a flexbox-only [LayoutConfig] from the IR property pair list.
     *
     * @param properties (propertyType, data) pairs, usually from
     *                   IRComponent.properties.map { it.type to it.data }.
     * @return A LayoutConfig whose flexbox fields are populated from the
     *         input and whose grid/position/advanced/root fields are all null.
     */
    fun extract(properties: List<Pair<String, JsonElement?>>): LayoutConfig {
        // Accumulator starts as the identity scaffold and gets `.copy()`'d on
        // each hit. Using a var + copy keeps the type checker happy without a
        // builder helper class.
        var cfg = LayoutConfig.Empty

        properties.forEach { (type, data) ->
            // Keywords come through as uppercase primitives (e.g. "FLEX_START").
            // We lowercase + swap `_`→`-` so the match arms mirror CSS syntax
            // exactly — this keeps grep'ability against the CSS spec high.
            //
            // Guarded: extractKeyword calls .jsonPrimitive on the inner
            // "value" field, which throws IAE for object-shaped numeric data
            // (e.g. FlexGrow's {"value":{"type":"Number","value":1.0}}).
            // Swallow those so downstream branches that use `data` directly
            // (FlexGrow/FlexShrink/Order/FlexBasis) can still run.
            val kw = try {
                ValueExtractors.extractKeyword(data)
            } catch (e: IllegalArgumentException) {
                null
            }?.lowercase()?.replace('_', '-')

            cfg = when (type) {
                // --- display ------------------------------------------------
                // Mirrors DisplayProperty.kt value set. Unknown keywords fall
                // through to `Block` (CSS default for a non-flex element).
                "Display" -> cfg.copy(display = parseDisplay(kw))

                // --- flex axis ----------------------------------------------
                "FlexDirection" -> cfg.copy(flexDirection = parseFlexDirection(kw))
                "FlexWrap" -> cfg.copy(flexWrap = parseFlexWrap(kw))

                // Legacy -webkit-box-orient: vertical ≈ flex-direction: column.
                // Only applied when the explicit FlexDirection slot is still
                // empty, so an authored `flex-direction` always wins.
                "BoxOrient" -> if (cfg.flexDirection == null) {
                    cfg.copy(flexDirection = parseBoxOrient(kw))
                } else cfg

                // --- content + item alignment -------------------------------
                "JustifyContent" -> cfg.copy(justifyContent = parseAlignment(kw))
                "AlignItems" -> cfg.copy(alignItems = parseAlignment(kw))
                "AlignContent" -> cfg.copy(alignContent = parseAlignment(kw))
                "AlignSelf" -> cfg.copy(alignSelf = parseAlignment(kw))

                // --- numeric item properties --------------------------------
                // The IR wraps flex-grow/shrink as
                //   { "value": { "type": "...Number", "value": 1.5 } }
                // which ValueExtractors.extractFloat (one-level-deep) misses.
                // [deepFloat] / [deepInt] below walk the `value` chain. Null
                // results fall through cleanly so an unparseable value stays
                // null, NOT 0 (preserves the semantic of "unset").
                "FlexGrow" -> deepFloat(data)
                    ?.let { cfg.copy(flexGrow = it) } ?: cfg
                "FlexShrink" -> deepFloat(data)
                    ?.let { cfg.copy(flexShrink = it) } ?: cfg
                "Order" -> deepInt(data)
                    ?.let { cfg.copy(order = it) } ?: cfg

                // --- flex-basis ---------------------------------------------
                // FlexBasis has a compound grammar: keyword (auto/content) or
                // <length-percentage>. Step 1 LayoutConfig only exposes a
                // `Default` placeholder, so we map any non-null input to it;
                // the richer variants arrive when FlexBasisValue is expanded
                // in a later step. TODO(phase7): wire length/percentage cases
                // once FlexBasisValue has Length / Percentage / Auto variants.
                "FlexBasis" -> cfg.copy(flexBasis = FlexBasisValue.Default)

                // Anything else: pass through unchanged — the other
                // sub-extractors will claim it.
                else -> cfg
            }
        }

        return cfg
    }

    // --- keyword parsers -----------------------------------------------------
    //
    // Kept as private functions (not inline `when`s) so the test suite can
    // cover each mapping surface with a minimal fixture, and so adding a new
    // keyword is a one-line diff.

    /**
     * Walk a JSON value looking for the first primitive numeric payload,
     * descending through `{ "value": ... }` wrappers up to [maxDepth] levels.
     *
     * The IR commonly wraps scalars as
     *   {"value": {"type": "...Number", "value": 1.5}}
     * and ValueExtractors.extractFloat only unwraps the outermost level.
     * Rather than patch the shared extractor (risk of breaking unrelated
     * callers), we do the deep walk locally.
     */
    private fun deepFloat(json: JsonElement?, maxDepth: Int = 3): Float? {
        if (json == null) return null
        var node: JsonElement = json
        repeat(maxDepth) {
            when (node) {
                is JsonPrimitive -> return (node as JsonPrimitive).floatOrNull
                is JsonObject -> {
                    val obj = node as JsonObject
                    // Prefer "value" then "numeric" — matches ValueExtractors' order.
                    node = obj["value"] ?: obj["numeric"] ?: return null
                }
                else -> return null
            }
        }
        return (node as? JsonPrimitive)?.floatOrNull
    }

    /** Int analogue of [deepFloat]. */
    private fun deepInt(json: JsonElement?, maxDepth: Int = 3): Int? {
        if (json == null) return null
        var node: JsonElement = json
        repeat(maxDepth) {
            when (node) {
                is JsonPrimitive -> return (node as JsonPrimitive).intOrNull
                is JsonObject -> {
                    val obj = node as JsonObject
                    node = obj["value"] ?: obj["numeric"] ?: return null
                }
                else -> return null
            }
        }
        return (node as? JsonPrimitive)?.intOrNull
    }

    /** Map CSS `display` keyword → [DisplayKind]. Null/unknown → Block. */
    private fun parseDisplay(kw: String?): DisplayKind = when (kw) {
        "flex" -> DisplayKind.Flex
        "inline-flex" -> DisplayKind.InlineFlex
        "grid" -> DisplayKind.Grid
        "inline-grid" -> DisplayKind.InlineGrid
        "inline" -> DisplayKind.Inline
        "inline-block" -> DisplayKind.InlineBlock
        "none" -> DisplayKind.None
        "contents" -> DisplayKind.Contents
        "flow-root" -> DisplayKind.FlowRoot
        // Default: block. Mirrors legacy FlexExtractor.parseDisplayType.
        else -> DisplayKind.Block
    }

    /** Map CSS `flex-direction` keyword → [EngineFlexDirection]. */
    private fun parseFlexDirection(kw: String?): EngineFlexDirection = when (kw) {
        "row-reverse" -> EngineFlexDirection.RowReverse
        "column" -> EngineFlexDirection.Column
        "column-reverse" -> EngineFlexDirection.ColumnReverse
        else -> EngineFlexDirection.Row
    }

    /** Map CSS `flex-wrap` keyword → [EngineFlexWrap]. */
    private fun parseFlexWrap(kw: String?): EngineFlexWrap = when (kw) {
        "wrap" -> EngineFlexWrap.Wrap
        "wrap-reverse" -> EngineFlexWrap.WrapReverse
        else -> EngineFlexWrap.NoWrap
    }

    /**
     * Map legacy `-webkit-box-orient` → [EngineFlexDirection].
     *
     * Per spec: `horizontal` / `inline-axis` = row; `vertical` / `block-axis`
     * = column. Used only as a fallback when FlexDirection is absent.
     */
    private fun parseBoxOrient(kw: String?): EngineFlexDirection = when (kw) {
        "vertical", "block-axis" -> EngineFlexDirection.Column
        else -> EngineFlexDirection.Row
    }

    /**
     * Map a CSS `<content-alignment>` / `<self-alignment>` keyword → the
     * unified [AlignmentKeyword] enum. We deliberately accept the union of
     * both grammars here and rely on the Applier to reject combinations that
     * don't make sense (e.g. `stretch` for justify-content).
     */
    private fun parseAlignment(kw: String?): AlignmentKeyword = when (kw) {
        // Physical + logical edges collapse to Start/End because Compose
        // Arrangement maps them to the same values once LayoutDirection is
        // resolved by CompositionLocal.
        "start" -> AlignmentKeyword.Start
        "end" -> AlignmentKeyword.End
        "flex-start" -> AlignmentKeyword.FlexStart
        "flex-end" -> AlignmentKeyword.FlexEnd
        "center" -> AlignmentKeyword.Center
        "stretch" -> AlignmentKeyword.Stretch
        "space-between" -> AlignmentKeyword.SpaceBetween
        "space-around" -> AlignmentKeyword.SpaceAround
        "space-evenly" -> AlignmentKeyword.SpaceEvenly
        "baseline" -> AlignmentKeyword.Baseline
        "normal" -> AlignmentKeyword.Normal
        "auto" -> AlignmentKeyword.Auto
        // CSS Anchor Positioning Level 1 §6: `anchor-center` resolves to
        // `center` whenever no default anchor is in scope. SDUI has no
        // anchor-positioning runtime so this fold is always correct.
        // Mirrors the iOS mapAlignment + Web AlignSelfApplier behaviour.
        "anchor-center" -> AlignmentKeyword.Center
        // Null / unknown keyword: pick `Normal` since CSS `initial` value for
        // most align-* properties is `normal`. Callers treat Normal as
        // "use platform default" — matching Compose's defaults (Start/Top).
        else -> AlignmentKeyword.Normal
    }
}
